app/com/arpnetworking/metrics/portal/scheduling/impl/BaseSchedule.java

```diff
@@ -0,0 +1,137 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.portal.scheduling.impl;
+
+import com.arpnetworking.commons.builder.OvalBuilder;
+import com.arpnetworking.metrics.portal.scheduling.Schedule;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import net.sf.oval.constraint.NotNull;
+import net.sf.oval.constraint.ValidateWithMethod;
+
+import java.time.ZonedDateTime;
+import java.util.Optional;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * Abstract base class for {@link Schedule}s.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public abstract class BaseSchedule implements Schedule {
+
+    private final ZonedDateTime _runAtAndAfter;
+    private final Optional<ZonedDateTime> _runUntil;
+
+    /**
+     * Protected constructor.
+     *
+     * @param builder Instance of <code>Builder</code>.
+     */
+    protected BaseSchedule(final Builder<?, ?> builder) {
+        _runAtAndAfter = builder._runAtAndAfter;
+        _runUntil = builder._runUntil;
+    }
+
+    protected ZonedDateTime getRunAtAndAfter() {
+        return _runAtAndAfter;
+    }
+
+    protected Optional<ZonedDateTime> getRunUntil() {
+        return _runUntil;
+    }
+
+    @Override
+    public Optional<ZonedDateTime> nextRun(final Optional<ZonedDateTime> lastRun) {
+        Optional<ZonedDateTime> result = unboundedNextRun(lastRun);
+        while (result.isPresent() && result.get().isBefore(_runAtAndAfter)) {
+            result = unboundedNextRun(result);
+        }
+        if (_runUntil.isPresent() && result.isPresent() && result.get().isAfter(_runUntil.get())) {
+            return Optional.empty();
+        }
+        return result;
+    }
+
+    /**
+     * Return the next time the schedule should run, without regard for the bounds set by {@code runAtAndAfter}/{@code runUntil}.
+     * ({@code nextRun} wraps this method and enforces the window mask.)
+     *
+     * @param lastRun The last time the job was run.
+     * @return The next time to run the job.
+     */
+    protected abstract Optional<ZonedDateTime> unboundedNextRun(Optional<ZonedDateTime> lastRun);
+
+    /**
+     * Builder implementation for {@link BaseSchedule} subclasses.
+     *
+     * @param <B> type of the builder
+     * @param <S> type of the object to be built
+     *
+     * @author Spencer Pearson (spencerpearson at dropbox dot com)
+     */
+    protected abstract static class Builder<B extends Builder<B, S>, S extends BaseSchedule> extends OvalBuilder<S> {
+        @NotNull
+        @ValidateWithMethod(methodName = "validateRunAtAndAfter", parameterType = ZonedDateTime.class)
+        protected ZonedDateTime _runAtAndAfter;
+        @NotNull
+        protected Optional<ZonedDateTime> _runUntil = Optional.empty();
```

Comment:
By convention the member fields on the Builder are not enclosed in `Optional` but can be marked `@Nullable`. The wrapping occurs when the fields are interpreted by the constructor.

Reply:
Sorry, I missed that last time.